openai_sdk_agent_app_sample_code/cookbook/agents/function_calling_agent.py

```diff
@@ -116,62 +116,72 @@ def predict(
         )
 
         # Call the LLM to recursively calls tools and eventually deliver a generation to send back to the user
-        (
-            model_response,
-            messages_log_with_tool_calls,
-        ) = self.recursively_call_and_run_tools(messages=messages)
-
-        # If your front end keeps of converastion history and automatically appends the bot's response to the messages history, remove this line.
-        messages_log_with_tool_calls.append(
-            model_response.choices[0].message.to_dict()
-        )  # OpenAI client
+        messages_log_with_tool_calls = self.recursively_call_and_run_tools(
+            messages=messages
+        )
 
         # remove the system prompt - this should not be exposed to the Agent caller
         messages_log_with_tool_calls = messages_log_with_tool_calls[1:]
 
         return {
-            "content": model_response.choices[0].message.content,
+            # "content": model_response.choices[0].message.content,
+            "content": messages_log_with_tool_calls[-1]["content"],
             # messages should be returned back to the Review App (or any other front end app) and stored there so it can be passed back to this stateless agent with the next turns of converastion.
             "messages": messages_log_with_tool_calls,
         }
 
-    @mlflow.trace(span_type="AGENT")
+    @mlflow.trace(span_type="CHAIN")
```

Comment:
QQ, why change from agent to chain?

Reply:
Mostly asking since this does still seem like the core agentic loop logic

Reply:
(same thinking as below)